File: bongf/java/src/ch16_dp_questions/Q32_IntegerTriangle.java
package ch16_dp_questions;

import java.util.Arrays;
import java.util.Scanner;

//https://www.acmicpc.net/problem/1932
public class Q32_IntegerTriangle {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        sc.nextLine();
        Integer[] now = Arrays.stream(sc.nextLine().split(" ")).map(Integer::valueOf).toArray(Integer[]::new);
        Integer[] prev = now;

        for (int i = 1; i < n; i++) {
            now = Arrays.stream(sc.nextLine().split(" ")).map(Integer::valueOf).toArray(Integer[]::new);
            for (int j = 0; j < now.length; j++) {
                if (j == 0) {
                    now[j] = prev[0] + now[j];
                    continue;
                }
                if (j == now.length - 1) {
                    now[j] = prev[prev.length - 1] + now[j];
                    continue;
                }
                now[j] = Math.max(prev[j - 1], prev[j]) + now[j];
            }
            prev = now;
        }

        int ans = Arrays.stream(now).mapToInt(x -> x).max().getAsInt();
        System.out.println(ans);
    }
}

File: bongf/python/dongbinbook/ch16_dp_questions/q32_integer_triangle.py
## https://www.acmicpc.net/problem/1932
def solution() :
    n = int(input())
    now = list(map(int, input().split(" ")))
    prev = now

    for i in range(1, n):
        now = list(map(int, input().split(" ")))
        for j in range(len(now)) :
            if(j==0):
                now[j] = prev[0] + now[j]
            elif (j==len(now)-1):
                now[j] = prev[-1] + now[j]
            else :
                now[j] = max(prev[j-1], prev[j]) + now[j]
        prev = now
    
    return max(now)

print(solution())

        